src/logger.rs

```diff
@@ -0,0 +1,44 @@
+// Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
+
+use crocksdb_ffi;
+use libc::{c_char, c_void};
+use librocksdb_sys::{DBEnv, DBLogger};
+use std::ffi::{CString, VaList};
+
+pub trait Logger: Send + Sync {
+    fn logv(&self, format: *const c_char, ap: VaList);
+}
+
+#[allow(dead_code)]
+extern "C" fn destructor(ctx: *mut c_void) {
+    unsafe {
+        Box::from_raw(ctx as *mut Box<dyn Logger>);
+    }
+}
+
+#[allow(dead_code)]
+extern "C" fn logv(ctx: *mut c_void, format: *const c_char, ap: VaList) {
+    unsafe {
+        let logger = &*(ctx as *mut Box<dyn Logger>);
+        logger.logv(format, ap);
+    }
+}
+
+#[allow(dead_code)]
+pub fn new_logger<L: Logger>(l: L) -> *mut DBLogger {
+    let p: Box<dyn Logger> = Box::new(l);
+    unsafe {
+        let logger_impl = crocksdb_ffi::crocksdb_logger_impl_create(
+            Box::into_raw(Box::new(p)) as *mut c_void,
+            destructor,
+            logv,
+        );
+        crocksdb_ffi::crocksdb_logger_create_from_impl(logger_impl)
+    }
+}
+
+#[allow(dead_code)]
+pub fn create_env_logger(fname: &str, mut env: DBEnv) -> *mut DBLogger {
+    let name = CString::new(fname.as_bytes()).unwrap();
+    unsafe { crocksdb_ffi::crocksdb_create_env_logger(name.as_ptr(), &mut env) }
+}
```

Comment:
I want to make this API look pretty. Can we make `create_env_logger` return a rust struct that implements `Logger` trait. To do that you need to implement the corresponding `crocksdb_logger_logv` function.

Reply:
Why do we need to add an ```Envlogger``` struct? I think that we have created the ``EnvLogger` object by the ```create_env_logger```. it implemented the ```logv``` in the rocksdb repo.

Reply:
I would like to make log format be the same as vanilla rocksdb, but some of the formatting logic is in `EnvLogger` (e.g. the datetime format: https://github.com/facebook/rocksdb/blob/master/logging/env_logger.h#L107). One way to do is to have the tikv `Logger` only handling file rotation logic, and creates `EnvLogger` for actually log writing. Alternatively we can duplicate the logic from `EnvLogger` into the tikv `Logger` to get the same log format. I prefer the former but it is up to you.

Reply:
.I think duplicate the logic from EnvLogger into the tikv is a good idea and keey log format style like tikv. rust-rocksdb is a library. It should be opened to change the logger for users. users can make decision to how to implement logger.For example, tikv can write rocksdb log to disk, can upload to hdfs. and have foramt you want.

Reply:
Compromise is that I can write a example to show how to complement it.

Reply:
@hawkingrei make sense. I'll merge the code as is. I didn't concern about user of don't know how to implement a logger. I was wanting to have the tikv logger being consistent with original rocksdb log format. But you are right it should consistent with tikv log format instead.